### PR TITLE
Added changes to support validation of security roles for plugins

### PR DESCRIPTION
**Pull request by saikaranam-amazon**
##  opensearch-security pull request intake form
_Please provide as much details as possible to get feedback/acceptance on your PR quickly_

 
 1. __Category:__ (_Enhancement, New feature, Bug fix, Test fix, Refactoring, Maintenance, Documentation_)
 New feature
 
 
 2. __Github Issue # or road-map entry, if available:__
https://github.com/opensearch-project/security/issues/1337

 3. __Description of changes:__
Added assume role support for the plugins - #1337 

 4. __Why these changes are required?__ 
These changes are required for picking-up specified roles associated with an user.

 5. __What is the old behavior before changes and new behavior after changes?__ (_Please add any example/logs/screen-shot if available_)

 6. __Testing done:__ (_Please provide details of testing done: Unit testing, integration testing and manual testing_)
 Added Integration tests and tested locally
 
 
 7. __TO-DOs, if any:__ (_Please describe pending items and provide Github issues# for each of them_)
N/A

 8. __Is it backport from main branch?__ (_If yes, please add backport PR # and commits #_)
N/A

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by vengadanathan-s — APPROVED** [review]

**Comment by saikaranam-amazon** [comment]
@hardik-k-shah @cliu123 Addressed the comments. Please review and merge the PR

File: src/main/java/org/opensearch/security/filter/SecurityFilter.java [hunk 1]
```diff
@@ -348,9 +348,13 @@ public void onFailure(Exception e) {
                 }
             } else {
                 auditLog.logMissingPrivileges(action, request, task);
-                String err = (injectedRoles == null) ?
-                        String.format("no permissions for %s and %s", pres.getMissingPrivileges(), user) :
-                        String.format("no permissions for %s and associated roles %s", pres.getMissingPrivileges(), injectedRoles);
+                String err = String.format("no permissions for %s and %s", pres.getMissingPrivileges(), user);
+                if(!pres.getMissingSecurityRoles().isEmpty()) {
+                    err = String.format("no permissions for %s to assume roles %s", user, pres.getMissingSecurityRoles());
```
Comment (hardik-k-shah) [inline src/main/java/org/opensearch/security/filter/SecurityFilter.java:353]: Update error message to remove assume role context.

File: src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java [hunk 2]
```diff
@@ -206,14 +208,26 @@ public PrivilegesEvaluatorResponse evaluate(final User user, String action0, fin
             action0 = PutMappingAction.NAME;
         }
 
+        final PrivilegesEvaluatorResponse presponse = new PrivilegesEvaluatorResponse();
+
         final TransportAddress caller = threadContext.getTransient(ConfigConstants.OPENDISTRO_SECURITY_REMOTE_ADDRESS);
-        final Set<String> mappedRoles = (injectedRoles == null) ? mapRoles(user, caller) : injectedRoles;
+        Set<String> mappedRoles = (injectedRoles == null) ? mapRoles(user, caller) : injectedRoles;
+        final String assumeRolesString = threadContext.getTransient(ConfigConstants.OPENDISTRO_SECURITY_ROLES_VALIDATION);
```
Comment (hardik-k-shah) [inline src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java:215]: Please remove assume role references from everywhere (file name, constants, variable, error message and comments) .

File: src/main/java/org/opensearch/security/filter/SecurityFilter.java (hunk 1 above)
Comment (saikaranam-amazon) [reply src/main/java/org/opensearch/security/filter/SecurityFilter.java]: Done

File: src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java [hunk 3]
```diff
@@ -206,14 +208,26 @@ public PrivilegesEvaluatorResponse evaluate(final User user, String action0, fin
             action0 = PutMappingAction.NAME;
         }
 
+        final PrivilegesEvaluatorResponse presponse = new PrivilegesEvaluatorResponse();
+
         final TransportAddress caller = threadContext.getTransient(ConfigConstants.OPENDISTRO_SECURITY_REMOTE_ADDRESS);
-        final Set<String> mappedRoles = (injectedRoles == null) ? mapRoles(user, caller) : injectedRoles;
+        Set<String> mappedRoles = (injectedRoles == null) ? mapRoles(user, caller) : injectedRoles;
+        final String rolesValidationString = threadContext.getTransient(ConfigConstants.OPENDISTRO_SECURITY_ROLES_VALIDATION);
```
[the remark this answers is not in the document]
Comment (saikaranam-amazon) [reply src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java]: Done

File: src/main/java/org/opensearch/security/support/ConfigConstants.java [hunk 4]
```diff
@@ -265,6 +265,10 @@
     // Roles injection for plugins
     public static final String OPENDISTRO_SECURITY_INJECTED_ROLES = "opendistro_security_injected_roles";
 
+    // Roles validation for the plugins
+    public static final String OPENDISTRO_SECURITY_ROLES_VALIDATION = "opendistro_security_roles_validation";
```
[the remark this answers is not in the document]
Comment (saikaranam-amazon) [reply src/main/java/org/opensearch/security/support/ConfigConstants.java]: Done

File: src/main/java/org/opensearch/security/filter/SecurityFilter.java [hunk 5]
```diff
@@ -348,9 +348,13 @@ public void onFailure(Exception e) {
                 }
             } else {
                 auditLog.logMissingPrivileges(action, request, task);
-                String err = (injectedRoles == null) ?
-                        String.format("no permissions for %s and %s", pres.getMissingPrivileges(), user) :
-                        String.format("no permissions for %s and associated roles %s", pres.getMissingPrivileges(), injectedRoles);
+                String err = String.format("no permissions for %s and %s", pres.getMissingPrivileges(), user);
```
[the remark this answers is not in the document]
Comment (saikaranam-amazon) [reply src/main/java/org/opensearch/security/filter/SecurityFilter.java]: Done

**Review by hardik-k-shah — DISMISSED** [review]

File: src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (saikaranam-amazon) [reply src/main/java/org/opensearch/security/privileges/PrivilegesEvaluator.java]: Adding the reference here for the validation checks for replication: [replication-plugin-ref](https://github.com/opensearch-project/cross-cluster-replication/blob/main/src/main/kotlin/org/opensearch/replication/action/setup/TransportSetupChecksAction.kt#L157-L167)

**Review by cliu123 — APPROVED** [review]

**Comment by saikaranam-amazon** [comment]
The Tests failed unrelated to this commit in `RolesInjectorIntegTest` due to test issue where security plugin is not initialized before running the tests.

**Review by cliu123 — DISMISSED** [review]
Changes look good to me. Triggered CI. Please merge once CI passes.

**Review by hardik-k-shah — APPROVED** [review]

**Review by lukkoor — APPROVED** [review]

**Pull request merged**